File: tests/crash_test_scalability.py
import sys
import os
import time
import requests
import subprocess
import psutil
import csv
import json
import statistics
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed

# --- CONFIG ---
ROOT_DIR = Path(__file__).parent.parent.resolve()
START_SCRIPT = ROOT_DIR / "start_llm.py"
from test_utils import _default_models_dir
_MODELS_DIR = _default_models_dir()
MODEL_PATH = _MODELS_DIR / "qwen2.5-0.5b-instruct-q5_k_m.gguf"
REPORT_FILE = ROOT_DIR / "scalability_report.csv"
START_PORT = 9000


class ScalabilityTester:
    """ScalabilityTester class."""

    def __init__(self, max_instances=10, safe_mode=False):
        """Initialize instance."""
        self.procs = []
        self.metrics = []
        self.max_instances = max_instances
        self.safe_mode = safe_mode
        self.active_ports = []

    def log(self, msg):
        # [X-Ray auto-fix] print(f"[CrashTest] {msg}")
        pass

    def clean_all(self):
        """Clean all."""
        self.log("Cleaning up processes...")
        for p in self.procs:
            try:
                p.terminate()
            except Exception:
                pass

        # Aggressive cleanup of orphaned llama-server
        for p in psutil.process_iter(["pid", "name"]):
            if p.info["name"] != "llama-server.exe":
                continue
            try:
                p.terminate()
            except Exception:
                pass
        self.procs = []
        self.active_ports = []
        time.sleep(2)

    def launch_instance(self, port):
        """Launch instance."""
        env = os.environ.copy()
        env["LLM_PORT"] = str(port)

        # Guard Bypass is Essential
        cmd = [sys.executable, str(START_SCRIPT), "--model", str(MODEL_PATH), "--guard-bypass"]

        proc = subprocess.Popen(
            cmd,
            cwd=ROOT_DIR,
            stdout=subprocess.DEVNULL,  # Mute stdout for cleaner benchmark logs
            stderr=subprocess.DEVNULL,
            env=env,
            stdin=subprocess.PIPE,
            shell=False,
        )
        return proc

    def wait_health(self, port, timeout_sec=60):
        """Wait health."""
        start = time.time()
        while time.time() - start < timeout_sec:
            try:
                r = requests.get(f"http://127.0.0.1:{port}/health", timeout=1)
                if r.status_code == 200:
                    return True
            except Exception:
                time.sleep(1)
        return False

    def benchmark_single(self, port, question="Verify system check."):
        """Measure inference time on a single instance"""
        start_t = time.time()
        end_t = 0

        # Simple completion payload
        payload = {
            "prompt": f"<|im_start|>user\n{question}<|im_end|>\n<|im_start|>assistant\n",
            "n_predict": 128,
            "temperature": 0.7,
            "stream": False,  # Streaming is harder to measure simply without client, we measure total duration
        }

        try:
            # We measure Total Round Trip Time (RTT) here as proxy for load
            # Proper TTFT requires streaming support
            resp = requests.post(f"http://127.0.0.1:{port}/completion", json=payload, timeout=60)
            end_t = time.time()
            if resp.status_code == 200:
                return {
                    "success": True,
                    "duration": end_t - start_t,
                    "tokens": resp.json().get("timings", {}).get("predicted_n", 0),
                    "port": port,
                }
        except Exception:
            pass

        return {"success": False, "port": port}


def _do_do_run_cycle_setup_setup(count):
    """Helper: setup phase for _do_run_cycle_setup."""

    self.log(f"--- Starting Cycle N={count} ---")
    cycle_ports = [START_PORT + i for i in range(count)]

    # Launch Group
    launch_start = time.time()
    for p in cycle_ports:
        proc = self.launch_instance(p)
        self.procs.append(proc)

    # Wait for Ready
    ready_count = 0
    for p in cycle_ports:
        if self.wait_health(p):
            ready_count += 1
        else:
            self.log(f"Port {p} failed to start")

    launch_duration = time.time() - launch_start
    self.log(f"Ready: {ready_count}/{count} (Launch took {launch_duration:.2f}s)")

    if ready_count < count:
        self.log("CRITICAL: Not all instances started. System limit reached?")
        return False

    return cycle_ports, i, launch_duration, p

    return cycle_ports, i, launch_duration, p


def _do_run_cycle_setup_part1():
    """Do run cycle setup part 1."""

    def run(self):
        """Run."""
        # CSV Header
        with open(REPORT_FILE, "w", newline="") as f:
            writer = csv.DictWriter(
                f, fieldnames=["instances", "success_rate", "avg_latency", "launch_overhead", "cpu_usage", "ram_usage"]
            )
            writer.writeheader()

        for n in range(1, self.max_instances + 1):
            if self.run_cycle(n):
                continue
            self.log("Stopping due to failure.")
            break

            # Append Record
            with open(REPORT_FILE, "a", newline="") as f:
                writer = csv.DictWriter(
                    f,
                    fieldnames=[
                        "instances",
                        "success_rate",
                        "avg_latency",
                        "launch_overhead",
                        "cpu_usage",
                        "ram_usage",
                    ],
                )
                writer.writerow(self.metrics[-1])

            self.clean_all()
            if self.safe_mode and n >= 2:
                self.log("Safe Mode: Stopping at N=2")
                break

        self.log(f"Test Complete. Report saved to {REPORT_FILE}")


def _do_run_cycle_setup(count):
    """Helper: setup phase for run_cycle."""
    cycle_ports, i, launch_duration, p = _do_do_run_cycle_setup_setup(count)

    def run_cycle(self, count):
        """Run cycle."""
        cycle_ports, i, launch_duration, p = _do_run_cycle_setup(count)
        # Benchmark Parallel
        self.log("Generating Parallel Load...")
        questions = [
            "Explain quantum physics roughly.",
            "Write a poem about rust.",
            "Calculate fibonacci 10.",
            "Describe the solar system.",
            "Why is the sky blue?",
            "How does a CPU work?",
            "What is DNA?",
            "History of Rome.",
        ]

        results = []
        with ThreadPoolExecutor(max_workers=count) as executor:
            futures = []
            for i, p in enumerate(cycle_ports):
                q = questions[i % len(questions)]
                futures.append(executor.submit(self.benchmark_single, p, q))

            for ft in as_completed(futures):
                results.append(ft.result())

        # Analyze
        successes = [r for r in results if r["success"]]
        avg_time = statistics.mean([r["duration"] for r in successes]) if successes else 0
        total_tokens = sum([r.get("tokens", 0) for r in successes])
        total_tokens / avg_time if avg_time > 0 else 0  # Rough estimate per instance stream

        row = {
            "instances": count,
            "success_rate": len(successes) / count,
            "avg_latency": avg_time,
            "launch_overhead": launch_duration,
            "cpu_usage": psutil.cpu_percent(),
            "ram_usage": psutil.virtual_memory().percent,
        }
        self.metrics.append(row)
        self.log(f"Cycle Result: {json.dumps(row, indent=2)}")

        return len(successes) == count

    _do_run_cycle_setup_part1()


if __name__ == "__main__":
    # If run directly with no args, default to safe mode (N=2) for verification
    # User can run with "python tests/crash_test_scalability.py FULL" for full test
    is_safe = True
    if len(sys.argv) > 1 and sys.argv[1] == "FULL":
        is_safe = False

    tester = ScalabilityTester(max_instances=10, safe_mode=is_safe)
    try:
        tester.clean_all()  # Pre-clean
        tester.run()
    finally:
        tester.clean_all()

File: rust_output/src/crash_test_scalability.rs
use anyhow::{Result, Context};
use crate::test_utils::{_default_models_dir};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};

pub static ROOT_DIR: std::sync::LazyLock<String /* Path.parent().unwrap_or(std::path::Path::new("")).parent().unwrap_or(std::path::Path::new("")).resolve */> = std::sync::LazyLock::new(|| Default::default());

pub const START_SCRIPT: &str = "ROOT_DIR / 'start_llm::py";

pub static _MODELS_DIR: std::sync::LazyLock<_default_models_dir> = std::sync::LazyLock::new(|| Default::default());

pub const MODEL_PATH: &str = "_MODELS_DIR / 'qwen2.5-0.5b-instruct-q5_k_m.gguf";

pub const REPORT_FILE: &str = "ROOT_DIR / 'scalability_report.csv";

pub const START_PORT: i64 = 9000;

/// ScalabilityTester class.
#[derive(Debug, Clone)]
pub struct ScalabilityTester {
    pub procs: Vec<serde_json::Value>,
    pub metrics: Vec<serde_json::Value>,
    pub max_instances: String,
    pub safe_mode: String,
    pub active_ports: Vec<serde_json::Value>,
}

impl ScalabilityTester {
    /// Initialize instance.
    pub fn new(max_instances: String, safe_mode: String) -> Self {
        Self {
            procs: vec![],
            metrics: vec![],
            max_instances,
            safe_mode,
            active_ports: vec![],
        }
    }
    pub fn log(&self, msg: String) -> () {
        println!("[CrashTest] {}", msg);
        // pass
    }
    /// Clean all.
    pub fn clean_all(&mut self) -> Result<()> {
        // Clean all.
        self.log("Cleaning up processes...".to_string());
        for p in self.procs.iter() {
            // try:
            {
                p.terminate();
            }
            // except Exception as _e:
        }
        for p in psutil.process_iter(vec!["pid".to_string(), "name".to_string()]).iter() {
            if p.info["name".to_string()] != "llama-server::exe".to_string() {
                continue;
            }
            // try:
            {
                p.terminate();
            }
            // except Exception as _e:
        }
        self.procs = vec![];
        self.active_ports = vec![];
        Ok(std::thread::sleep(std::time::Duration::from_secs_f64(2)))
    }
    /// Launch instance.
    pub fn launch_instance(&self, port: String) -> Result<()> {
        // Launch instance.
        let mut env = os::environ.clone();
        env["LLM_PORT".to_string()] = port.to_string();
        let mut cmd = vec![sys::executable, START_SCRIPT.to_string(), "--model".to_string(), MODEL_PATH.to_string(), "--guard-bypass".to_string()];
        let mut proc = subprocess::Popen(cmd, /* cwd= */ ROOT_DIR, /* stdout= */ subprocess::DEVNULL, /* stderr= */ subprocess::DEVNULL, /* env= */ env, /* stdin= */ subprocess::PIPE, /* shell= */ false);
        Ok(proc)
    }
    /// Wait health.
    pub fn wait_health(&self, port: String, timeout_sec: String) -> Result<()> {
        // Wait health.
        let mut start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
        while (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start) < timeout_sec {
            // try:
            {
                let mut r = /* reqwest::get( */&format!("http://127.0.0.1:{}/health", port)).cloned().unwrap_or(/* timeout= */ 1);
                if r.status_code == 200 {
                    true
                }
            }
            // except Exception as _e:
        }
        Ok(false)
    }
    /// Measure inference time on a single instance
    pub fn benchmark_single(&self, port: String, question: String) -> Result<()> {
        // Measure inference time on a single instance
        let mut start_t = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
        let mut end_t = 0;
        let mut payload = HashMap::from([("prompt".to_string(), format!("<|im_start|>user\n{}<|im_end|>\n<|im_start|>assistant\n", question)), ("n_predict".to_string(), 128), ("temperature".to_string(), 0.7_f64), ("stream".to_string(), false)]);
        // try:
        {
            let mut resp = /* reqwest::post( */format!("http://127.0.0.1:{}/completion", port), /* json= */ payload, /* timeout= */ 60);
            let mut end_t = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
            if resp.status_code == 200 {
                HashMap::from([("success".to_string(), true), ("duration".to_string(), (end_t - start_t)), ("tokens".to_string(), resp.json().get(&"timings".to_string()).cloned().unwrap_or(HashMap::new()).get(&"predicted_n".to_string()).cloned().unwrap_or(0)), ("port".to_string(), port)])
            }
        }
        // except Exception as _e:
        Ok(HashMap::from([("success".to_string(), false), ("port".to_string(), port)]))
    }
}

/// Helper: setup phase for _do_run_cycle_setup.
pub fn _do_do_run_cycle_setup_setup(count: String) -> () {
    // Helper: setup phase for _do_run_cycle_setup.
    self.log(format!("--- Starting Cycle N={} ---", count));
    let mut cycle_ports = 0..count.iter().map(|i| (START_PORT + i)).collect::<Vec<_>>();
    let mut launch_start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    for p in cycle_ports.iter() {
        let mut proc = self.launch_instance(p);
        self.procs.push(proc);
    }
    let mut ready_count = 0;
    for p in cycle_ports.iter() {
        if self.wait_health(p) {
            ready_count += 1;
        } else {
            self.log(format!("Port {} failed to start", p));
        }
    }
    let mut launch_duration = (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - launch_start);
    self.log(format!("Ready: {}/{} (Launch took {:.2}s)", ready_count, count, launch_duration));
    if ready_count < count {
        self.log("CRITICAL: Not all instances started. System limit reached?".to_string());
        false
    }
    (cycle_ports, i, launch_duration, p)
    (cycle_ports, i, launch_duration, p)
}

/// Do run cycle setup part 1.
pub fn _do_run_cycle_setup_part1() -> Result<()> {
    // Do run cycle setup part 1.
    let run = || {
        // Run.
        let mut f = File::create(REPORT_FILE)?;
        {
            let mut writer = csv::DictWriter(f, /* fieldnames= */ vec!["instances".to_string(), "success_rate".to_string(), "avg_latency".to_string(), "launch_overhead".to_string(), "cpu_usage".to_string(), "ram_usage".to_string()]);
            writer.writeheader();
        }
        for n in 1..(self.max_instances + 1).iter() {
            if self.run_cycle(n) {
                continue;
            }
            self.log("Stopping due to failure.".to_string());
            break;
            let mut f = File::create(REPORT_FILE)?;
            {
                let mut writer = csv::DictWriter(f, /* fieldnames= */ vec!["instances".to_string(), "success_rate".to_string(), "avg_latency".to_string(), "launch_overhead".to_string(), "cpu_usage".to_string(), "ram_usage".to_string()]);
                writer.writerow(self.metrics[-1]);
            }
            self.clean_all();
            if (self.safe_mode && n >= 2) {
                self.log("Safe Mode: Stopping at N=2".to_string());
                break;
            }
        }
        self.log(format!("Test Complete. Report saved to {}", REPORT_FILE));
    Ok(})
}

/// Helper: setup phase for run_cycle.
pub fn _do_run_cycle_setup(count: String) -> () {
    // Helper: setup phase for run_cycle.
    let (mut cycle_ports, mut i, mut launch_duration, mut p) = _do_do_run_cycle_setup_setup(count);
    let run_cycle = |count| {
        // Run cycle.
        let (mut cycle_ports, mut i, mut launch_duration, mut p) = _do_run_cycle_setup(count);
        self.log("Generating Parallel Load...".to_string());
        let mut questions = vec!["Explain quantum physics roughly.".to_string(), "Write a poem about rust.".to_string(), "Calculate fibonacci 10.".to_string(), "Describe the solar system.".to_string(), "Why is the sky blue?".to_string(), "How does a CPU work?".to_string(), "What is DNA?".to_string(), "History of Rome.".to_string()];
        let mut results = vec![];
        let mut executor = ThreadPoolExecutor(/* max_workers= */ count);
        {
            let mut futures = vec![];
            for (i, p) in cycle_ports.iter().enumerate().iter() {
                let mut q = questions[(i % questions.len())];
                futures.push(executor.submit(self.benchmark_single, p, q));
            }
            for ft in as_completed(futures).iter() {
                results.push(ft.result());
            }
        }
        let mut successes = results.iter().filter(|r| r["success".to_string()]).map(|r| r).collect::<Vec<_>>();
        let mut avg_time = if successes { statistics.mean(successes.iter().map(|r| r["duration".to_string()]).collect::<Vec<_>>()) } else { 0 };
        let mut total_tokens = successes.iter().map(|r| r.get(&"tokens".to_string()).cloned().unwrap_or(0)).collect::<Vec<_>>().iter().sum::<i64>();
        if avg_time > 0 { (total_tokens / avg_time) } else { 0 };
        let mut row = HashMap::from([("instances".to_string(), count), ("success_rate".to_string(), (successes.len() / count)), ("avg_latency".to_string(), avg_time), ("launch_overhead".to_string(), launch_duration), ("cpu_usage".to_string(), psutil.cpu_percent()), ("ram_usage".to_string(), psutil.virtual_memory().percent)]);
        self.metrics.push(row);
        self.log(format!("Cycle Result: {}", serde_json::to_string(&row).unwrap()));
        successes.len() == count
    };
    _do_run_cycle_setup_part1();
}
